docs/developer/design-v2-2023-11-11.md

```diff
@@ -0,0 +1,198 @@
+
+```hcl
+provider "materialize" {
+  app_password   = var.materialize_app_password
+  default_region = var.materialize_default_region
+  endpoint       = var.materialize_endpoint
+}
+```
+
+Users will specify the app password and the default region for operations. Optionally, the endpoint can be overridden for testing or staging purposes.
+
+### Multi-Client Structure
+
+The provider will maintain a map of clients for database operations across different regions and a client for interfacing with the Frontegg API.
+
+```hcl
+type ProviderMeta struct {
+    DB       map[Region]*clients.DBClient
+    Frontegg *clients.FronteggClient
+    // Future: CloudAPI *clients.CloudAPIClient
+}
+```
+
+This structure allows for scalability and extension of the provider to incorporate additional clients as necessary.
+
+## Resources and Operations
+
+### Region Resource
+
+Responsible for managing the availability of Materialize regions. It will interact with the Cloud API to enable or retrieve regions as requested by the user.
+
+```hcl
+resource "materialize_region" "aws_us_east_1" {
+  region = "aws/us-east-1"
+}
+```
+
+Operations:
+
+-   **Enable Region**: Uses the Cloud API to enable a new region.
+-   **Read Region**: Retrieves the status of a region from the Cloud API.
+-   **Disable Region**: Noop. Regions cannot be disabled once enabled.
+
+### User Resource
+
+Manages the lifecycle of a user within the Frontegg platform.
+
+```hcl
+resource "materialize_app_user" "example_user" {
+  email          = "user@example.com"
+  auth_provider  = "local"
+  role_ids       = ["role-id-1", "role-id-2"] # Logic to retrieve role IDs from role names?
+  # Additional attributes...
+}
+```
+
+Operations:
+
+-   **Create User**: Invokes the Frontegg API to create/invite a user.
+-   **Read User**: Retrieves user details from Frontegg.
+-   **Update User**: Attaches or detaches roles from the user.
+-   **Delete User**: Removes the user from Frontegg.
+
+### Organization Data Source?
+
+Generates a list of available organizations from Frontegg.
+
+```hcl
+data "materialize_app_organizations" "all" {}
+```
+
+Operations:
+
+-   **List Organizations**: Fetches all available organizations from Frontegg.
+
+### Role Data Source
+
+Generates a list of available roles from Frontegg.
+
+```hcl
+data "materialize_app_roles" "all" {}
+```
+
+Operations:
+
+-   **List Roles**: Fetches all available roles from Frontegg.
+
+## Testing Strategy
+
+Along the existing unit tests described in the [design document](../../DESIGN.md), the provider will include a comprehensive suite of tests for all resources and operations, including:
+
+-   **Unit Tests**: Will run against a test Materialize environment or a mock service to ensure that the provider functions correctly in isolation.
+-   **Integration Tests**: Will run against staging instance to confirm that the provider interacts correctly with the actual services.
+-   **Acceptance Tests**: То be determined.
```

Comment:
One thing to be careful of here: I think CI currently runs integration tests against a database Docker container. Once you start using the cloud API, you're not going to be able to use the Docker image alone to run integration tests, because that database Docker container won't have the cloud APIs.

There are a few options:

  * @mjibson is working on a mock Frontegg implementation you could use to mock out Frontegg. (That won't help you with the region enable API though; you'd have to write your own mock for that.)
  * You could follow the console repo's approach: clone the cloud repo in CI and start the local development environment, which provides a testing Frontegg environment, a local cloud API, and a local Materialize environment.
  * You could always run CI against staging.

I'm honestly not sure which is best.

Reply:
Yeah, the mocks you added in this PR look good!